Inject neutral plasma and grid refactor

roelof-groenewald opened the pull request:
This PR got away from me a little so there is more stuff in here than ideally would have been.
Volume injector and emitters were added to allow us to seed the plasma with a net-neutral plasma. I then moved the simulation grid declaration into `mwxrun.py` to allow us to get the grid properties before WarpX is initiated. This in turn triggered a refactor of the diode setup script which then required all the unit tests to be updated with the new format.
The only test that had reference values change was `test_direct_solver` which previously used the builtin WarpX functionality to seed a plasma. This has been changed to use our own volumetric injector (which places particles randomly rather than on a neat grid). The overall change in the reference results were very minimal as can be seen in the atteched plot.
![direct_solver_change](https://user-images.githubusercontent.com/40245517/126720019-d6f6f528-42f5-4d57-a4fe-dd4dcc6c0bf0.png)

This should be merged after PR #28 .

nruof committed:
Cherry-Pick from nruof/get_grid_quantities to access proc number from mwxrun

nruof committed:
Additional changes to make capacitive_discharge.py to run on this branch

nruof committed:
Add files from thermionic emission branch from ME_WarpX

nruof committed:
Create thermionic_emission.py example

nruof committed:
Merge branch 'development' of https://github.com/ModernElectron/WarpX into nruof/thermionic_emission

nruof committed:
Edits to thermionic_emission.py to make it run without errors

nruof committed:
In principle thermionic_ignition script, edit to Evolve so ElectroStaticSolver uses particle injection, typo in fields.py

nruof committed:
Merge branch 'memaster2' of https://github.com/ModernElectron/WarpX into nruof/thermionic_emission

nruof committed:
Merge branch 'memaster' of https://github.com/ModernElectron/WarpX into nruof/thermionic_emission

nruof committed:
Merge fix, multiple definitions for warpx_getMyProc, warpx_getNProcs

nruof committed:
Merge branch 'memaster' of https://github.com/ModernElectron/WarpX into nruof/thermionic_emission

nruof committed:
addition merge from capacitive_discharge.py

nruof committed:
WarpXWrapper changes from merge

nruof committed:
Merge branch 'memaster' of https://github.com/ModernElectron/WarpX into nruof/thermionic_emission

nruof committed:
Thermionic emission example with long simulation time

nruof committed:
Merge branch 'memaster' of https://github.com/ModernElectron/WarpX into nruof/thermionic_emission

nruof committed:
Updates to use MCC wrapper, mwxconstants, and mepicmi.Species adds to simulation by default

nruof committed:
Removed minerva from util.py

nruof committed:
Remove dependency to minerva

nruof committed:
End of line in mwxconstants

nruof committed:
Apply suggestions from code review, spacing changes

Co-authored-by: peterscherpelz <31747262+peterscherpelz@users.noreply.github.com>

nruof committed:
Removed unused var in thermionic_ignition, change iproc nproc to point to mwxrun, edited assemblies.py doc string

nruof committed:
Doc string updates in emission.py

nruof committed:
Update mewarpx/mewarpx/emission.py, spacing

Co-authored-by: peterscherpelz <31747262+peterscherpelz@users.noreply.github.com>

nruof committed:
Update mewarpx/mewarpx/emission.py, doc string

Co-authored-by: peterscherpelz <31747262+peterscherpelz@users.noreply.github.com>

nruof committed:
Spacing

Co-authored-by: peterscherpelz <31747262+peterscherpelz@users.noreply.github.com>

roelof-groenewald committed:
exposes AddRealComp to Python to allow extra particle attributes to be added at runtime; also includes a new function to grab a particle data array from the name of the component rather than the index

roelof-groenewald committed:
exposes AddRealComp to Python to allow extra particle attributes to be added at runtime; also includes a new function to grab a particle data array from the name of the component rather than the index

roelof-groenewald committed:
added wrappers to add new PIDs at runtime and grab particle data given a PID name

nruof committed:
Setup test for thermionic_emission

nruof committed:
Merge branch 'memaster' of https://github.com/ModernElectron/WarpX into nruof/thermionic_emission

nruof committed:
Unit Test for thermionic emission that compares net charge density

nruof committed:
test_thermionic_emission comparison file

nruof committed:
Formatting: Revert WarpXEvolve, mepicmi DocString, use picmi constants, update warnings

nruof committed:
Apply suggestions from code review, spacing

Co-authored-by: peterscherpelz <31747262+peterscherpelz@users.noreply.github.com>
Co-authored-by: Kevin Zhu <86268612+KZhu-ME@users.noreply.github.com>
Co-authored-by: Roelof Groenewald <40245517+roelof-groenewald@users.noreply.github.com>

nruof committed:
Edits to emission.py, Spacing and f-strings

nruof committed:
Merge branch 'nruof/thermionic_emission' of https://github.com/ModernElectron/WarpX into nruof/thermionic_emission

nruof committed:
Bug / Syntax fixes

roelof-groenewald committed:
added functionality to AddNParticles() to allow extra particle attributes to also be set during particle creation

nruof committed:
Set USE_SCHOTTKY to False by default and raise error if set to True since no implementation yet

nruof committed:
Match WarpXEvolve to upstream PR

roelof-groenewald committed:
added function to get index of a particle component given the PID name

roelof-groenewald committed:
changed new get component index and get_particle_arrays_from_comp_name functions to take species name as argument rather than species id

roelof-groenewald committed:
changed warpx_addRealComp to accept a species name as input and only add the new component for that species

nruof committed:
Doc String updates, spacing

nruof committed:
Apply suggestions from code review, spacing

Co-authored-by: Peter Scherpelz <31747262+peterscherpelz@users.noreply.github.com>
Co-authored-by: Roelof Groenewald <40245517+roelof-groenewald@users.noreply.github.com>

nruof committed:
Merge branch 'nruof/thermionic_emission' of https://github.com/ModernElectron/WarpX into nruof/thermionic_emission

roelof-groenewald committed:
Merge branch 'python_runtime_particle_comp' into add_pids_runtime_python

roelof-groenewald committed:
Merge branch 'addNParticles_extra_comps' into add_pids_runtime_python

roelof-groenewald committed:
update mwxrun functions to use species name rather than id to add a new PID or access particle data

nruof committed:
f strings, small edits

roelof-groenewald committed:
Merge remote-tracking branch 'public/development' into add_pids_runtime_python

roelof-groenewald committed:
Merge remote-tracking branch 'public/nruof/thermionic_emission' into add_pids_runtime_python

roelof-groenewald committed:
initial fixes for new _libwarpx functions

roelof-groenewald committed:
fixed failing direct solver test

roelof-groenewald committed:
added tolerance to direct solver result comparisons since it seems that different systems give slightly different results (tested on my workstation and on penning), also changed it to only compare phi since that is what we care about in this test

roelof-groenewald committed:
Merge branch 'nruof/thermionic_emission' into add_pids_runtime_python

roelof-groenewald committed:
fixed bug in add_particles to correctly determine the number of extra attributes

roelof-groenewald committed:
updated test_custom_criteria.py for new libwarpx functions

roelof-groenewald committed:
Merge remote-tracking branch 'public/memaster' into add_pids_runtime_python

roelof-groenewald committed:
added functionality to mewarpx to add new particle data arrays at runtime

roelof-groenewald committed:
changed the thermionic_ignition.py example to demonstrate the functionality of the new PIDs

roelof-groenewald committed:
added a test for a runtime pid - added this to the test of terminating from a custom criteria

roelof-groenewald committed:
Merge remote-tracking branch 'public/memaster' into add_pids_runtime_python

roelof-groenewald committed:
fixed bug in AddNParticles if fewer attribute values are passed than the number of extra arrays for the species

roelof-groenewald committed:
added volumetric injector and Cartesian volume emitter

roelof-groenewald committed:
refactored diode setup and current tests with the new grid declaration approach

roelof-groenewald committed:
Merge remote-tracking branch 'public/memaster' into inject_neutral_plasma_and_grid_refactor and update circle emitter test (also moved it into test_emission.py)

roelof-groenewald committed:
added a test for the plasma injector

KZhu-ME commented on mewarpx/mewarpx/setups_store/diode_setup.py:
```diff
@@ -380,38 +383,37 @@ def init_solver(self):
-            V=self.V_ANODE, z=self.D_CA,
-            zsign=1, WF=self.ANODE_PHI, T=self.ANODE_TEMP,
-            name='anode_plane'
+        self.anode_plane = assemblies.Anode(
```
I think it would help with consistency to have both cathode and anode variable names to follow the same format, for example, naming them `cathode_plane` and `anode_plane`.

peterscherpelz commented on mewarpx/mewarpx/emission.py:
```diff
@@ -418,6 +410,219 @@ def inject_particles(self):
+        """Initialize injection of a plasma with two species and given emitter.
+
+        Arguments:
+            emitter (:class:`mewarpx.emission.Emitter`): BaseEmitter object
```
```suggestion
            emitter (:class:`mewarpx.emission.BaseEmitter`): BaseEmitter object
```

peterscherpelz commented on mewarpx/mewarpx/emission.py:
```diff
@@ -418,6 +410,219 @@ def inject_particles(self):
+                that will specify positions and velocities of particles to
+                inject.
+            species1 (mepicmi.Species): First species, eg electron
+            species2 (mepicmi.Species): Second species, eg ion
```
```suggestion
            species1 (:class:`mewarpx.mepicmi.Species`): First species, eg electron
            species2 (:class:`mewarpx.mepicmi.Species`): Second species, eg ion
```

peterscherpelz commented on mewarpx/mewarpx/emission.py:
```diff
@@ -1059,3 +1264,128 @@ def get_normals(self, x, y, z):
         normals[:, 2] = self.normal[self.contour_idx, 1]
         return normals
+
+
```
```suggestion
```

peterscherpelz reviewed (commented):
I'll do a final review when #28 is merged. The one thing I wonder about is whether to go to the effort of adding a surface emitter to the unit test (or as a 2nd unit test to separate it out), or to disable it in the code if we're not testing it (or at least add a warning it's untested?) or something...

I also had separate unit tests for volume emitter emission statistics, and for when T2 is fixed. Is there a reason you dropped those? I'm not sure your test covers all the cases tested there.

roelof-groenewald committed:
Apply suggestions from code review by PS

Co-authored-by: Peter Scherpelz <31747262+peterscherpelz@users.noreply.github.com>

roelof-groenewald committed:
added PlasmaInjector tests for a surface emitter and a case where T_2 is specified

roelof-groenewald committed:
Merge remote-tracking branch 'public/memaster' into inject_neutral_plasma_and_grid_refactor

roelof-groenewald committed:
code cleanup

 Please enter the commit message for your changes. Lines starting

PhilMiller commented on mewarpx/tests/test_direct_solver.py:
```diff
@@ -82,6 +83,9 @@ def test_direct_solver():
     # uncomment to generate reference data
     # np.save('reference_data.npy', data)
 
+    # the MGML solver has the last data point set to 0
```
Typo:  MLMG

PhilMiller commented on mewarpx/tests/test_direct_solver.py:
```diff
@@ -38,33 +38,34 @@ def test_direct_solver():
-            lambda t: VOLTAGE * np.sin(2.0 * np.pi * FREQ * t) if DIRECT_SOLVER
-            else "%.1f*sin(2*pi*%.5e*t)" % (VOLTAGE, FREQ)
+            (lambda t: VOLTAGE * np.sin(2.0 * np.pi * FREQ * t))
+            if DIRECT_SOLVER else f"{VOLTAGE}*sin(2*pi*{FREQ:.5e}*t)"
```
Once the parser is available from Python, this will be able to go away, right?

roelof-groenewald replied:
Yeap, that is the hope / plan.

peterscherpelz committed:
Fix a few inconsistencies in grid and emission.

Also fixed a number of formatting issues that came up, a few of which
are in other parts of the affected files.

peterscherpelz reviewed (approved):
@roelof-groenewald I just submitted a commit with my changes, which fixed a number of formatting warnings I saw and a few unrelated comments. Catching an rseed reset in CartesianVolumeEmitter and a few self.grid misses were the only substantive changes I believe.

peterscherpelz commented:
@KZhu-ME @MKieburtz @nruof Can one of you please do a review on this PR before merge? Thanks!

EtherealMC-Bit commented on mewarpx/mewarpx/mwxconstants.py:
```diff
@@ -8,6 +8,7 @@
 e = constants.q_e # C
 m_e = constants.m_e # kg
 kb_J = constants.kb # J/K
+epsilon_0 = constants.ep0
```
Add units here maybe?

EtherealMC-Bit reviewed (approved):
Looks good! Just one comment

KZhu-ME reviewed (approved):
Looks good!

roelof-groenewald committed:
added units to the declaration of epsilon0 in mwxconstants.py
